tests/failpoints/cases/test_pending_peers.rs

```diff
@@ -225,3 +225,37 @@ fn test_on_check_busy_on_apply_peers() {
     let stats = cluster.pd_client.get_store_stats(3).unwrap();
     assert!(!stats.is_busy);
 }
+
+#[test]
+fn test_on_apply_snap_failed() {
+    let mut cluster = new_node_cluster(0, 3);
+    cluster.cfg.raft_store.raft_base_tick_interval = ReadableDuration::millis(5);
+    cluster.cfg.raft_store.raft_store_max_leader_lease = ReadableDuration::millis(100);
+    cluster.cfg.raft_store.pd_heartbeat_tick_interval = ReadableDuration::millis(100);
+
+    let pd_client = Arc::clone(&cluster.pd_client);
+    // Disable default max peer count check.
+    pd_client.disable_default_operator();
+
+    let region_id = cluster.run_conf_change();
+    pd_client.must_add_peer(region_id, new_peer(2, 2));
+
+    // To ensure peer 2 is not pending.
+    cluster.must_put(b"k1", b"v1");
+    must_get_equal(&cluster.get_engine(2), b"k1", b"v1");
+
+    // Mock applying snapshot failed on peer 3.
+    fail::cfg("region_apply_snap_io_err", "return").unwrap();
+    pd_client.must_add_peer(region_id, new_peer(3, 3));
+    sleep_ms(100);
```

Comment:
Better replace this kind of sleep with `test_util::eventually` to avoid flaky cases.

Reply:
Not appropriate for this, it should wait one region_heartbeat for updating the meta.